test/integration/webhook_test.go

```diff
@@ -169,6 +182,64 @@ func TestValidationWebhook(t *testing.T) {
 	//       See: https://github.com/Kong/kubernetes-ingress-controller/issues/1442
 	time.Sleep(time.Second * 5)
 
+	t.Log("creating a gatewayclass to verify gateway validation")
+	gatewayc, err := gatewayclient.NewForConfig(env.Cluster().Config())
+	require.NoError(t, err)
+	gatewayClass := &gatewayv1alpha2.GatewayClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: uuid.NewString(),
+		},
+		Spec: gatewayv1alpha2.GatewayClassSpec{
+			ControllerName: gateway.ControllerName,
+		},
+	}
+	gatewayClass, err = gatewayc.GatewayV1alpha2().GatewayClasses().Create(ctx, gatewayClass, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	defer func() {
+		t.Logf("cleaning up gatewayclass %s", gatewayClass.Name)
+		assert.NoError(t, gatewayc.GatewayV1alpha2().GatewayClasses().Delete(ctx, gatewayClass.Name, metav1.DeleteOptions{}))
+	}()
+
+	t.Log("creating an invalid gateway to verify that validation fails")
+	gateway := &gatewayv1alpha2.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "kong",
+			// the missing annotations here make the gateway invalid
+		},
+		Spec: gatewayv1alpha2.GatewaySpec{
+			GatewayClassName: gatewayv1alpha2.ObjectName(gatewayClass.Name),
+			Listeners: []gatewayv1alpha2.Listener{{
+				Name:     "http",
+				Protocol: gatewayv1alpha2.HTTPProtocolType,
+				Port:     gatewayv1alpha2.PortNumber(80),
+			}},
+		},
+	}
+	_, err = gatewayc.GatewayV1alpha2().Gateways(ns.Name).Create(ctx, gateway, metav1.CreateOptions{})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "missing required annotation")
+
+	t.Log("verifying that we don't validate a gateway that belongs to another controller")
```

Comment:
wouldn't refactoring this into a table-based test help readability and ease addition of further test cases?

Reply:
Sure, I'll change it to a loop

Reply:
So I'm not actually sure, because I tested the logic paths with a _single_ object in this case so my thinking is making it into a loop where I suddenly have several objects would be a lot more code and ultimately might _reduce_ readability? LMKYT

Reply:
We now have a test that validates unrelated bits of functionality (secret validation and gateway validation) within a single test that is 700 lines long, has the flows for individual isolatable tests mixed, and has unclear patterns of reuse. I'm worried about maintainability of such a test I'd like to look into that (and maybe the nondeterminism coming from the `sleep`) in a PR to this PR I've just started drafting. Stay tuned :radio:

Reply:
https://github.com/Kong/kubernetes-ingress-controller/pull/2051

> the testing suite is due for a refactor where admission webhook setup is part of environment setup and not part of specific test setup such that multiple tests can separately validate that functionality

#2051 makes it possible for admission webhook setup to happen in the context of a specific test case, which (I think) is better than setting up as part of the environment, because it puts the filtering logic closer to the validating logic, and does not bloat the test harness further. The approach proposed in that PR makes it possible to have multiple tests.

I think that we're at the point where the amount of technical debt accrued in the integration tests is too much. Flakiness is a real problem; I view simplifying (and isolating) test cases to be a good countermeasure.

Reply:
> because it puts the filtering logic closer to the validating logic, and does not bloat the test harness further.

I don't see globally configuring the admission webhook as bloat, I instead see it as additional verification of functionality. We ask end-users to have it on at all times to catch mistakes and we can just as easily make mistakes in tests that might pass if the webhook were off but still technically be invalid or wrong in some way.

Reply:
> I think that we're at the point where the amount of technical debt accrued in the integration tests is too much

This I agree with

>  Flakiness is a real problem; I view simplifying (and isolating) test cases to be a good countermeasure.

Have you seen specific test flakes occur with this test? While I agree it needs work I haven't actually seen it be flaky for webhook testing specific reasons it appears to be consistently passing.

Reply:
I'll review #2051 and we can reconvene here after that.

Reply:
Alright I'm done with my review of #2051 and I think we can move forward with that. I appreciate the willingness to jump in and make a PR to make the changes you wanted to see.